src/computability/encoding.lean

```diff
@@ -0,0 +1,189 @@
+/-
+Copyright (c) 2020 Pim Spelier, Daan van Gent. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Pim Spelier, Daan van Gent.
+-/
+
+import data.fintype.basic
+import data.num.lemmas
+import tactic
+
+/-!
+# Encodings
+
+This file contains the definition of a (finite) encoding, a map from a type to
+strings in an alphabet, used in defining computability by Turing machines.
+It also contains several examples:
+
+## Examples
+
+- `fin_encoding_nat_Γ₀₁`   : a binary encoding of ℕ in a simple alphabet.
+- `fin_encoding_nat_Γ'`    : a binary encoding of ℕ in the alphabet used for TM's.
+- `unary_fin_encoding_nat` : a unary encoding of ℕ
+- `fin_encoding_bool_Γ₀₁`  : an encoding of bool.
+-/
+
+namespace computability
+
+/-- An encoding of a type in a certain alphabet, together with a decoding. -/
+structure encoding (α : Type) :=
+(Γ : Type)
+(encode : α → list Γ)
+(decode : list Γ → option α)
+(decode_encode : ∀ x, decode (encode x) = some x)
+
+/-- An encoding plus a guarantee of finiteness of the alphabet. -/
+structure fin_encoding (α : Type) extends encoding α :=
+(Γ_fin : fintype Γ)
+
+/-- An alphabet consisting of bit0 and bit1 -/
+@[derive [inhabited,decidable_eq,fintype]]
+inductive Γ₀₁
+| bit0 | bit1
+
+/-- A standard Turing machine alphabet, consisting of blank,bit0,bit1,bra,ket,comma. -/
+@[derive [decidable_eq,fintype]]
+inductive Γ'
+| blank | bit0 | bit1 | bra | ket | comma
```

Comment:
How about using `| bit (b : bool)` instead of `bit0` and `bit1`, so that the `inclusion` map is built in.